File: src/main/java/fi/finlit/edith/sql/domain/Paragraph.java
package fi.finlit.edith.sql.domain;

import java.util.ArrayList;
import java.util.List;

import javax.xml.stream.XMLStreamException;

import org.apache.commons.lang.StringUtils;

import fi.finlit.edith.util.ParagraphParser;

public class Paragraph {
    private final List<ParagraphElement> elements = new ArrayList<ParagraphElement>();

    public List<ParagraphElement> getElements() {
        return elements;
    }

    public Paragraph addElement(ParagraphElement e) {
        elements.add(e);
        return this;
    }

    public Paragraph copy() {
        Paragraph paragraph = new Paragraph();
        for (ParagraphElement element : elements) {
            paragraph.addElement(element.copy());
        }
        return paragraph;
    }

    @Override
    public int hashCode(){
        return toString().hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this){
            return true;
        }else if (obj instanceof Paragraph){
            return obj.toString().equals(toString());
        }else{
            return false;
        }
    }

    @Override
    public String toString() {
        return StringUtils.join(elements, "");
    }
}

File: src/main/scala/fi/finlit/edith/sql/domain/Paragraph.scala
package fi.finlit.edith.sql.domain

import java.util.ArrayList
import java.util.List
import org.apache.commons.lang.StringUtils
import scala.reflect.BeanProperty
//remove if not needed
import scala.collection.JavaConversions._

class Paragraph {

  @BeanProperty
  val elements = new ArrayList[ParagraphElement]()

  def addElement(e: ParagraphElement): Paragraph = {
    elements.add(e)
    this
  }

  override def hashCode(): Int = toString.hashCode

  override def equals(obj: Any): Boolean = {
    if (obj == this) {
      true
    } else if (obj.isInstanceOf[Paragraph]) {
      obj.toString == toString
    } else {
      false
    }
  }

  override def toString(): String = StringUtils.join(elements, "")
}
